core/base/composition.cpp

```diff
@@ -0,0 +1,101 @@
+/*******************************<GINKGO LICENSE>******************************
+Copyright 2017-2018
+
+Karlsruhe Institute of Technology
+Universitat Jaume I
+University of Tennessee
+
+Redistribution and use in source and binary forms, with or without modification,
+are permitted provided that the following conditions are met:
+
+1. Redistributions of source code must retain the above copyright notice,
+   this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright notice,
+   this list of conditions and the following disclaimer in the documentation
+   and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its contributors
+   may be used to endorse or promote products derived from this software
+   without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR
+ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include "core/base/composition.hpp"
+
+
+#include "core/matrix/dense.hpp"
+
+
+namespace gko {
+namespace {
+
+
+template <typename ValueType, typename OpIterator, typename VecIterator>
+inline void allocate_vectors(OpIterator begin, OpIterator end, VecIterator res)
+{
+    for (auto it = begin; it != end; ++it, ++res) {
+        if (*res != nullptr && (*res)->get_size()[0] == (*it)->get_size()[0]) {
+            continue;
+        }
+        *res = matrix::Dense<ValueType>::create(
+            (*it)->get_executor(), gko::dim<2>{(*it)->get_size()[0], 1});
+    }
+}
+
+
+inline const LinOp *apply_inner_operators(
+    const std::vector<std::shared_ptr<const LinOp>> &operators,
```

Comment:
I am wondering whether we should make this accessible from the outside. There are cases, where you want to turn around the order of the application - first outer then inner. Would exposing "apply_inner_operators" to the outside help in this case?

Reply:
I don't think so, this does something different - it applies all but the first operator, which is just an implementation detail to avoid code duplication.

For the use case you're mentioning you could decompose the operators, and then compose them again in reverse order. Where exactly do you have this situation?

In any case, there's one thing missing from the interface of both classes - getting the operators back. I'll create a separate PR for that when we merge this.

Reply:
``For the use case you're mentioning you could decompose the operators, and then compose them again in reverse order. Where exactly do you have this situation?``

For any solvers working with the transpose matrix, like QMR, for example.

Reply:
Well that's not exactly it. Applying the operator `A * B` in reverse would be `y = B(A(x))`. There, you would need `y = B^T(A^T(x))` which is just `(A * B)^T(x)`. We should just implement the transposable interface on `Combination` and `Composition` and problem solved - the solvers do not even have to know they're working with a decomposed operator.